attach/cid.c

```diff
@@ -0,0 +1,244 @@
+ */
+void cid_to_filename(struct Buffer *filename, struct CidMapList *cid_map_list)
+{
+  if (!filename || !cid_map_list)
+    return;
+
+  FILE *fp_out = NULL;
+  char *pbuf = NULL;
+  char *searchbuf = NULL;
+  char *buf = NULL;
+  char *cid = NULL;
+  size_t blen = 0;
+  struct CidMap *cid_map = NULL;
+
+  struct Buffer *tmpfile = mutt_buffer_pool_get();
+  struct Buffer *tmpbuf = mutt_buffer_pool_get();
+
+  FILE *fp_in = mutt_file_fopen(mutt_buffer_string(filename), "r");
+  if (!fp_in)
+    goto bail;
+
+  /* ensure tmpfile has the same file extension as filename otherwise an
+   * HTML file may be opened as plain text by the viewer */
+  const char *suffix = mutt_strn_rfind(mutt_buffer_string(filename),
+                                       mutt_buffer_len(filename), ".");
+  if (suffix && *(suffix++))
+    mutt_buffer_mktemp_pfx_sfx(tmpfile, "neomutt", suffix);
+  else
+    mutt_buffer_mktemp(tmpfile);
+  fp_out = mutt_file_fopen(mutt_buffer_string(tmpfile), "w+");
+  if (!fp_out)
+    goto bail;
+
+  /* Read in lines from filename in buf */
+  while ((buf = mutt_file_read_line(buf, &blen, fp_in, NULL, MUTT_RL_NO_FLAGS)) != NULL)
+  {
+    if (mutt_str_len(buf) == 0)
+    {
+      fputs(buf, fp_out);
+      continue;
+    }
+
+    /* copy buf to searchbuf because we need to edit multiple times */
+    searchbuf = mutt_str_dup(buf);
+    mutt_buffer_reset(tmpbuf);
+
+    /* loop through Content-ID to filename mappings and do search and replace */
+    STAILQ_FOREACH(cid_map, cid_map_list, entries)
+    {
+      pbuf = searchbuf;
+      while ((cid = strstr(pbuf, cid_map->cid)) != NULL)
+      {
+        mutt_buffer_addstr_n(tmpbuf, pbuf, cid - pbuf);
+        mutt_buffer_addstr(tmpbuf, cid_map->fname);
+        pbuf = cid + mutt_str_len(cid_map->cid);
+        mutt_debug(LL_DEBUG2, "replaced \"%s\" with \"%s\" in file \"%s\"\n",
+                   cid_map->cid, cid_map->fname, mutt_buffer_string(filename));
+      }
+      mutt_buffer_addstr(tmpbuf, pbuf);
+      FREE(&searchbuf);
+      searchbuf = mutt_buffer_strdup(tmpbuf);
+      mutt_buffer_reset(tmpbuf);
+    }
+
+    /* write edited line to output file */
+    fputs(searchbuf, fp_out);
+    fputs("\n", fp_out);
+    FREE(&searchbuf);
+  }
+
+  FREE(&buf);
+  mutt_file_fclose(&fp_in);
+  mutt_file_fclose(&fp_out);
+  mutt_file_set_mtime(mutt_buffer_string(filename), mutt_buffer_string(tmpfile));
+
+  /* add filename to TempAtachmentsList so it doesn't get left lying around */
+  mutt_add_temp_attachment(mutt_buffer_string(filename));
+  /* update filename to point to new file */
+  mutt_buffer_copy(filename, tmpfile);
+
+bail:
+
+  mutt_buffer_pool_release(&tmpfile);
+  mutt_buffer_pool_release(&tmpbuf);
+
+  /* free Content-ID to filename mapping list */
+  while (!STAILQ_EMPTY(cid_map_list))
+  {
+    cid_map = STAILQ_FIRST(cid_map_list);
+    STAILQ_REMOVE_HEAD(cid_map_list, entries);
+    cid_map_free(&cid_map);
+  }
```

Comment:
Also, I don't think this belongs here.

You create `cid_map_list` in `cid_save_attachments()`
It gets populated with a call to `cid_save_attachments()`
The contents get used, here in `cid_to_filename()`
I think the freeing should happen in `mutt_view_attachment()`
(and we can make the param const)